projects/Ephemery-Besu-Implementation.md

```diff
@@ -0,0 +1,200 @@
+# Ephemery Testnet Implementation on Besu
+
+Enables an automatically reset public testnet embedded on Besu
+
+
+
+## Motivation
+
+[EIP-6916](https://eips.ethereum.org/EIPS/eip-6916) (Ephemery) is introducing an ephemeral testnet that automatically resets and can provide an alternative environment for short-term testing of applications, validators, and also breaking changes in client implementations. Implementing EIP-6916 natively inside Ethereum clients such as Besu avoids issues of long-running testnets, which suffer from state bloat, lack of testnet funds, or consensus issues. Periodically resetting the network back to genesis cleans the validator set and returns funds back to faucets while keeping the network reasonably small for easy bootstrapping. This way a wider range of options would be available for users to examine their applications.
+
+
+## Project description
+
+
+
+This project aims to complete the remaining implementation of EIP-6916 in the [Besu client](https://github.com/hyperledger/besu), focusing specifically on the database reset feature and Geth-style genesis parsing compatibility, along with enabling the bootnode command to read from a dynamic source.
+
+
+
+## Specification
+
+
+The overall specification is defined in the EIP-6916. Besides that, the implementation on the client should be tailored to Besu's architecture. In the following, the steps required to get to final development are pointed out. 
+
+
+#### Reset database implementation
+
+- DB reset detection: This is when the current timestamp passes the genesis timestamp plus the period. The `EphemeryGenesisUpdater` class inside the app module can be an option for it.
+
+
+- DB cleaning: The database should be refreshed after each genesis update. Besu is using RocksDB; the implementation of cleaning the database needs adding functionality inside the `RocksDBKeyValueStorageFactory` class inside the rocksdb module. 
+
+
+
+- Refresh genesis state: after each network iteration, the `GenesisState` class inside the ethereum/core module should be updated with the new genesis state for a dynamic reset network. 
+
+
+
+
+#### Enable --bootnodes to read enode URLs from a file or remote URL
+
+
+- Changing the logic of the --bootnode command to not just read from the command parameter but also read from a file. This will enable the Ephemery testnet implementation to dynamically download a file, such as [boot nodes,](https://ephemery.dev/latest/metadata/enodes.txt) from a fixed URL. The BesuCommand class should be extended for enabling this functionality.
+
+
+- This way the bootnodes are going to be set dynamically, so peer-to-peer network logic might need to be changed throughout the codebase. 
```

Comment:
Bootnodes are not dynamic, can stay hardcoded even for Ephemery but could be useful to be read from URL for testnets. 

One feature to add here - When reset happens, the node needs to maintain its public key, drop old peers and rediscover peers on the new network